worker/codalabworker/fsm.py

```diff
@@ -0,0 +1,64 @@
+import threading
+import time
+
+
+class State(object):
+    """
+    The base class for a state in a finite state machine (FSM).
+    """
+    def update(self):
+        """
+        The update method which is called as part of the primary run loop of a FSM.
+        :return: a tuple of a new state for the FSM and a list of output messages
+        """
+        return self.transition_noop()
+
+    def transition_noop(self):
+        """
+        Transition to use when the state machine should make no changes.
+        :return: this state
+        """
+        return self
+
+    def transition_done(self):
+        """
+        Transition to use when the state machine is done.
+        :return: the 'done' state which will cause the state machine to be complete
+        """
+        return None
+
+
+# TODO Make an implementation which uses a single thread to manage many FSMs
+class ThreadedFiniteStateMachine(object):
+    """
+    A FSM which will run on a separate thread.
+    """
+    def __init__(self, initial_state, sleep_time=0.1):
+        self._state = initial_state
+        self._sleep_time = sleep_time
+        # The thread for running this FSM
+        self._thread = threading.Thread(target=ThreadedFiniteStateMachine._run, args=[self])
+        self._should_run = True
```

Comment:
I notice elsewhere in the code base we're using locks around shared booleans like this. I don't think we need to with CPython; does codalab target other implementations?

Reply:
@percyliang do you know the answer to this?